subsys/bluetooth/services/bas/bas.c

```diff
@@ -0,0 +1,152 @@
+#include <stdbool.h>
+#include <zephyr/types.h>
+#include <zephyr/sys/util_macro.h>
+
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/bluetooth/uuid.h>
+#include <zephyr/bluetooth/services/bas.h>
+#include "bas_internal.h"
+
+#define LOG_LEVEL CONFIG_BT_BAS_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bas, CONFIG_BT_BAS_LOG_LEVEL);
+
+#define BT_BAS_IDX_BATT_LVL_STATUS_CHAR_VAL 6
+
+/* Notify/Indicate all connections */
+struct bt_gatt_indicate_params ind_params;
+
+static struct bt_bas_data bt_bas_data = {
+	.battery_level = 100U,
+};
+
+static void blvl_ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	ARG_UNUSED(attr);
+
+	bool notif_enabled = (value == BT_GATT_CCC_NOTIFY);
+
+	LOG_INF("BAS Notifications %s", notif_enabled ? "enabled" : "disabled");
+}
+
+static void blvl_status_ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	ARG_UNUSED(attr);
+
+	bool notif_enabled = (value == BT_GATT_CCC_NOTIFY);
+	bool ind_enabled = (value == BT_GATT_CCC_INDICATE);
+
+	LOG_INF("BAS Notifications %s", notif_enabled ? "enabled" : "disabled");
+	LOG_INF("BAS Indications %s", ind_enabled ? "enabled" : "disabled");
+}
+
+static ssize_t read_blvl(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			 uint16_t len, uint16_t offset)
+{
+	uint8_t lvl8 = bt_bas_data.battery_level;
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &lvl8, sizeof(lvl8));
+}
+
+static ssize_t read_blvl_status(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+				uint16_t len, uint16_t offset)
+{
+	const struct bt_bas_bls *status = attr->user_data;
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, status, sizeof(struct bt_bas_bls));
+}
+
+static void indicate_cb(struct bt_conn *conn, struct bt_gatt_indicate_params *params, uint8_t err)
+{
+	if (err != 0) {
+		LOG_INF("Indication failed with error %d\n", err);
+	} else {
+		LOG_INF("Indication sent successfully\n");
+	}
+}
+
+/* Constant values from the Assigned Numbers specification:
+ * https://www.bluetooth.com/wp-content/uploads/Files/Specification/Assigned_Numbers.pdf?id=89
+ */
+static const struct bt_gatt_cpf level_cpf = {
+	.format = 0x04, /* uint8 */
+	.exponent = 0x0,
+	.unit = 0x27AD,        /* Percentage */
+	.name_space = 0x01,    /* Bluetooth SIG */
+	.description = 0x0106, /* "main" */
+};
+
+BT_GATT_SERVICE_DEFINE(
+	bas, BT_GATT_PRIMARY_SERVICE(BT_UUID_BAS),
+	BT_GATT_CHARACTERISTIC(BT_UUID_BAS_BATTERY_LEVEL, BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY,
+			       BT_GATT_PERM_READ, read_blvl, NULL, &bt_bas_data.battery_level),
+	BT_GATT_CCC(blvl_ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+	BT_GATT_CPF(&level_cpf),
+	BT_GATT_CHARACTERISTIC(BT_UUID_BAS_BATTERY_LEVEL_STATUS,
+			       BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_READ,
+			       read_blvl_status, NULL, &bt_bas_data.level_status),
```

Comment:
So I think I understand the design choice here. 

You've followed the design of `read_blvl` where the pointer to the battery level was supplied. The pointer was never used though, and since BAS is a singleton, then we can just access it as a global state. 

This also means that `level_status` does not need to be provided here and could actually be allocated in `bas_bls.c` as a global variable. This would seemingly simplify a few things as we don't need the `bt_bas_data` struct nor `bt_bas_bls_init`, as everything related to the battery level status can just reside inside of `bas_bls.c`. The only thing that needs an internal API is the read callback function